hathor/util.py

```diff
@@ -274,3 +275,29 @@ def not_none(optional: Optional[_T], message: str = 'Unexpected `None`') -> _T:
     if optional is None:
         raise AssertionError(message)
     return optional
+
+
+class Random(PyRandom):
+    def geometric(self, p: float) -> int:
+        """Port of numpy.random.Generator.geometric sampling
+
+        It uses the Inverse Transform Sampling [1].
+        CDF(x) = 1 - (1 - p)**x
+        CDF^{-1}(x) = log(1 - x) / log(1 - p)
+        [1] https://en.wikipedia.org/wiki/Inverse_transform_sampling
+        """
+        return math.ceil(math.log(self.random()) / math.log(1 - p))
+
+    def ordered_choices(self, seq: Sequence[T], k: int) -> List[T]:
+        """Like self.choices but preserve orginal order.
+
+        For example, ordered_choices([1, 2, 3]) will never return [3, 2] only [2, 3] instead."""
+        chosen = self.choices(seq, k=k)
+        chosen.sort(key=lambda x: seq.index(x))
```

Comment:
It might be two problems: (i) repeated elements, and (ii) time complexity O(n^2 log(n)) if the sort algorithm does not cache the keys.

Reply:
Fixed. Thanks!